### PR TITLE
Lion optimizer higher-LR sweep on DrivAerML cross-dataset

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

TandemFoil already achieves SOTA with Lion at lr=1.25e-4. DrivAerML uses AdamW at lr=5e-4. AirfRANS uses AdamW at lr=7e-4. Lion is known to be effective at 3–10× lower LRs than AdamW for the same problem, but its sensitivity to WD and gradient clipping is different. We hypothesize that Lion can match or beat AdamW on DrivAerML and AirfRANS at lr in the range 2e-4 to 5e-4, and that WD and gradient clipping are critical regularizers for Lion on these larger-geometry datasets.

This sweep tests Lion across DrivAerML at 4 learning rates (2e-4, 3e-4, 4e-4, 5e-4) and one AirfRANS Lion run, plus ablations that isolate the contribution of WD vs. grad-clip on DrivAerML. GPU 0 runs the TandemFoil baseline as a sanity check.

## Instructions

No changes to `train.py` are required.

Use `--wandb_group lion-higher-lr` on all runs. Run all 8 GPUs in parallel.

**GPU 0 — TandemFoil baseline (sanity check)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb_group lion-higher-lr \
  --run-name lion-lr1.25e4-tf-baseline &
```

**GPU 1 — AirfRANS + Lion lr=3e-4**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer lion --lr 3e-4 \
  --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group lion-higher-lr \
  --run-name lion-lr3e4-af &
```

**GPU 2 — DrivAerML + Lion lr=2e-4 (full regularization)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 python train.py \
  --dataset drivaerml \
  --optimizer lion --lr 2e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group lion-higher-lr \
  --run-name lion-lr2e4-wd1e2-gc1-dam &
```

**GPU 3 — DrivAerML + Lion lr=3e-4 (full regularization)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=3 python train.py \
  --dataset drivaerml \
  --optimizer lion --lr 3e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group lion-higher-lr \
  --run-name lion-lr3e4-wd1e2-gc1-dam &
```

**GPU 4 — DrivAerML + Lion lr=4e-4 (full regularization)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=4 python train.py \
  --dataset drivaerml \
  --optimizer lion --lr 4e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group lion-higher-lr \
  --run-name lion-lr4e4-wd1e2-gc1-dam &
```

**GPU 5 — DrivAerML + Lion lr=5e-4 (full regularization)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=5 python train.py \
  --dataset drivaerml \
  --optimizer lion --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group lion-higher-lr \
  --run-name lion-lr5e4-wd1e2-gc1-dam &
```

**GPU 6 — DrivAerML + Lion lr=3e-4 (no weight decay ablation)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=6 python train.py \
  --dataset drivaerml \
  --optimizer lion --lr 3e-4 \
  --cosine-t-max 30 --grad-clip 1.0 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group lion-higher-lr \
  --run-name lion-lr3e4-gc1-nowd-dam &
```

**GPU 7 — DrivAerML + Lion lr=3e-4 (no grad-clip ablation)**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=7 python train.py \
  --dataset drivaerml \
  --optimizer lion --lr 3e-4 \
  --cosine-t-max 30 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group lion-higher-lr \
  --run-name lion-lr3e4-wd1e2-nogc-dam &
```

## Baseline

Current best metrics:

| Dataset | Metric | Value | PR |
|---------|--------|-------|----|
| DrivAerML | val_primary/surface_rel_l2_pct | 4.619% | #2691 |
| AirfRANS | val_primary/surface_mse | 0.001236 | #2828 |
| TandemFoil | val_primary/surface_pressure_mae | 30.10 | #2810 |

**DrivAerML baseline run**: PR #2691, epoch 256, 4L/512d/8H, AdamW lr=5e-4, T_max=30
```
cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

**AirfRANS baseline run**: PR #2828, epoch 349, 2L/256d/4H, AdamW lr=7e-4, gc=1.0, WD=1e-2, T_max=5
```
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999
```

**TandemFoil baseline run**: PR #2810, epoch 157, 3L/192d/3H, Lion lr=1.25e-4, gc=1.0, WD=1e-2, T_max=10
```
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nami:

## Results

### Summary

**Lion fails catastrophically on DrivAerML and AirfRANS.** All 6 DrivAerML runs eventually diverged (grad_norm → NaN), and the AirfRANS run plateaued at 440× worse than baseline. Only the TandemFoil sanity check performed well, confirming Lion's effectiveness is dataset-specific and does not transfer to larger-geometry problems at these LRs.

### TandemFoil — Sanity Check (GPU 0)

| Metric | Baseline (PR #2810) | This Run (best) | This Run (test) | Delta |
|--------|-------------------|-----------------|-----------------|-------|
| val_primary/surface_pressure_mae | 30.10 | **27.57** (ep 260) | 30.56 | val -8.4%, test +1.5% |

- W&B: `ovv4gl5b` — Lion lr=1.25e-4, gc=1.0, wd=1e-2, T_max=10, 3L/192d/3H
- Epochs: 282 (timeout)
- **Sanity check passed.** Best val beats baseline by 8.4%. Test slightly regressed (30.56 vs 30.10), likely because test eval happened at final epoch rather than best-val epoch. Confirms Lion works well on TandemFoil.
- Note: `test_geom_camber_cruise/volume_mae` reported Infinity — possible data issue in test split.

### AirfRANS (GPU 1)

| Metric | Baseline (PR #2828) | This Run (best) | This Run (test) | Delta |
|--------|-------------------|-----------------|-----------------|-------|
| val_primary/surface_mse | 0.001236 | 0.00210 (ep 94) | 0.5465 | best +69.7%, test 442× worse |

- W&B: `vvq7ncq9` — Lion lr=3e-4, gc=1.0, wd=1e-2, T_max=5, 2L/256d/4H
- Epochs: 662 (timeout)
- **Complete failure.** Lion reached best val 0.00210 at epoch 94 then slowly degraded. By epoch 200+ the loss plateaued at ~1.46 and surface_mse was stuck at ~0.544. Grad norms averaged ~95-100 (vs ~20 early on), indicating persistent gradient instability even with gc=1.0.

### DrivAerML — LR Sweep (GPUs 2-5)

| GPU | Config | Best Val (%) | Best Epoch | Final Val (%) | Test (%) | W&B |
|-----|--------|-------------|------------|---------------|----------|-----|
| 2 | Lion lr=2e-4, gc=1.0, wd=1e-2 | **7.74** | 86 | 71.46 | 73.61 | `zzkwvjqu` |
| 3 | Lion lr=3e-4, gc=1.0, wd=1e-2 | 26.40 | 13 | 184.17 | 180.41 | `zyqsp6e5` |
| 4 | Lion lr=4e-4, gc=1.0, wd=1e-2 | 22.53 | 17 | 214.40 | 227.50 | `cguhy2dv` |
| 5 | Lion lr=5e-4, gc=1.0, wd=1e-2 | 37.90 | 8 | 70.49 | 70.84 | `0tuv2msz` |

Baseline: **4.619%** (PR #2691, AdamW lr=5e-4)

- **All runs diverged.** lr=2e-4 was the best, reaching 7.74% at epoch 86 before diverging — still 67% worse than the AdamW baseline.
- Higher LRs diverged faster and more severely. lr=5e-4 peaked at epoch 8 then grad_norm exploded to 153 by epoch 10.
- All runs showed grad_norm → NaN in later training, indicating fundamental instability.

### DrivAerML — Ablations (GPUs 6-7)

| GPU | Config | Best Val (%) | Best Epoch | Final Val (%) | Test (%) | W&B |
|-----|--------|-------------|------------|---------------|----------|-----|
| 3 | lr=3e-4, gc=1.0, wd=1e-2 (reference) | 26.40 | 13 | 184.17 | 180.41 | `zyqsp6e5` |
| 6 | lr=3e-4, gc=1.0, **no wd** | 13.29 | 35 | ~1100 | — | `mvr3fe2k` |
| 7 | lr=3e-4, **no gc**, wd=1e-2 | 35.08 | 17 | 76.34 | 76.54 | `y85c252f` |

- **Grad-clip is critical.** Without it (GPU 7), grad_norm reached 226M and performance was consistently worst.
- **Weight decay ablation (GPU 6)** showed better early convergence (13.3% vs 26.4% with WD), but diverged catastrophically to ~1100% — the worst final outcome of any run. WD initially hurts Lion's convergence but is needed for long-term stability.
- Neither ablation improved over the full-regularization run.

### Commands Used

```bash
# GPU 0 — TandemFoil baseline
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --agent nami --wandb-group lion-higher-lr --wandb-name "nami/lion-lr1.25e4-tf-baseline"

# GPU 1 — AirfRANS Lion lr=3e-4
CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --airfrans-task full --optimizer lion --lr 3e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --agent nami --wandb-group lion-higher-lr --wandb-name "nami/lion-lr3e4-af"

# GPUs 2-5 — DrivAerML LR sweep (lr=2e-4, 3e-4, 4e-4, 5e-4)
CUDA_VISIBLE_DEVICES={2,3,4,5} python train.py --dataset drivaerml --optimizer lion --lr {2e-4,3e-4,4e-4,5e-4} --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent nami --wandb-group lion-higher-lr

# GPU 6 — DrivAerML no-WD ablation
CUDA_VISIBLE_DEVICES=6 python train.py --dataset drivaerml --optimizer lion --lr 3e-4 --cosine-t-max 30 --grad-clip 1.0 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent nami --wandb-group lion-higher-lr --wandb-name "nami/lion-lr3e4-gc1-nowd-dam"

# GPU 7 — DrivAerML no-GC ablation
CUDA_VISIBLE_DEVICES=7 python train.py --dataset drivaerml --optimizer lion --lr 3e-4 --cosine-t-max 30 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --agent nami --wandb-group lion-higher-lr --wandb-name "nami/lion-lr3e4-wd1e2-nogc-dam"
```

### What Happened

**Lion does not transfer from TandemFoil to DrivAerML or AirfRANS.** The hypothesis that Lion could match AdamW at 2-5e-4 LR on these datasets is rejected.

The core failure mode is **late-training divergence**: all DrivAerML runs converge initially (some reaching reasonable early metrics) but then grad_norms explode to NaN/Infinity, causing catastrophic loss divergence. This happens even with gc=1.0 and wd=1e-2. Lion's sign-based update rule appears fundamentally incompatible with the loss landscape of these larger, more complex geometries.

On AirfRANS, Lion at lr=3e-4 never found a good basin — it reached 0.00210 surface_mse at best (69% worse than AdamW at 0.001236) and then degraded steadily.

**Key insight**: Lion's sign(m) update provides uniform step sizes regardless of gradient magnitude. On simple geometries (TandemFoil), this is a regularizing feature. On complex 3D geometries (DrivAerML 500k+ surface meshes), it appears to prevent the optimizer from making the fine-grained, magnitude-aware updates needed for convergence. The cosine LR schedule's periodic warm-ups may amplify this, repeatedly pushing the optimizer into unstable regimes.

### Suggested Follow-ups

1. **Lion with much lower LR for DrivAerML** (e.g., 5e-5 to 1e-4) — the lr=2e-4 run showed the most promise, suggesting the optimal Lion LR for DrivAerML might be even lower
2. **Lion with linear decay instead of cosine** — eliminating the periodic warm-ups that seem to trigger divergence
3. **Lion with stronger grad-clip** (e.g., 0.5 or 0.3) — given the instability, more aggressive clipping might stabilize training
4. **Investigate the TandemFoil Infinity in test** — the `test_geom_camber_cruise/volume_mae: Infinity` warrants a bug check

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2896 because ADVISOR: Closing — Lion is definitively incompatible with DrivAerML and AirfRANS. Every DrivAerML Lion run diverged catastrophically; AirfRANS catastrophic at 0.00210 vs 0.000699 new baseline. Even TandemFoil (Lion's native territory) failed to beat the current 26.06 baseline at 27.57. Do not revisit Lion on DrivAerML/AirfRANS..
